src/lib/services/contentful/graphqlClient.ts

```diff
@@ -0,0 +1,63 @@
+type SpaceID = string;
+type Token = string;
+type APIPrefix = string;
+
+const defaultAPIPrefix: APIPrefix = "https://graphql.contentful.com/content/v1/spaces";
+
+type ClientOptions = {
+  spaceID: SpaceID;
+  token: Token;
+  apiPrefix?: APIPrefix;
+};
+
+const delim = "#" as const;
+
+type ClientKey = `${APIPrefix}${typeof delim}${SpaceID}${typeof delim}${Token}`;
+const getKeyFromOptions = ({
+  spaceID,
+  token,
+  apiPrefix = defaultAPIPrefix,
+}: ClientOptions): ClientKey => `${apiPrefix}${delim}${spaceID}${delim}${token}`;
+
+export type Client = {
+  options: ClientOptions;
+  key: ClientKey;
+  fetch: <T>(query: string) => Promise<T>;
+};
+
+const clients = new Map<ClientKey, Client>();
+
+const getClient = ({ spaceID, token, apiPrefix = defaultAPIPrefix }: ClientOptions): Client => {
+  const key = getKeyFromOptions({ spaceID, token });
+  const existingClient = clients.get(key);
+  if (existingClient) return existingClient;
+
+  const client = {
+    options: { spaceID, token, apiPrefix },
+    key,
+    async fetch<T>(query: string): Promise<T> {
+      const { spaceID, token, apiPrefix } = this.options;
+      const url = `${apiPrefix}/${spaceID}`;
+      const response = await fetch(url, {
+        method: "POST",
+        headers: {
+          "Content-Type": "application/json",
+          Authorization: `Bearer ${token}`,
+        },
+        body: JSON.stringify({ query }),
+      });
+      if (!response.ok) {
+        throw new Error(
+          `Got failed response with status code ${response.status} ${response.statusText}`
+        );
```

Comment:
How do errors like this get handled? Do we need a try / catch rather than an if / else on the response in a route like `[...dynamicRoute]/+page.server.ts`?

Reply:
This will cause a 500 currently, but the follow-up PR to add server-side previewing also addresses this.